File: API/PetCenterModels/DataTransferObjects/Listing/ListingResponseDTO.cs
using Microsoft.VisualBasic;
using PetCenterModels.DBTables;
using PetCenterModels.DataTransferObjects;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Text.Json.Serialization;
using System.Threading.Tasks;
using PetCenterModels.SearchObjects;

namespace PetCenterModels.DataTransferObjects
{
    public class ListingResponseDTO : IAlbumCarryingDTO<Listing,ListingResponseDTO,ImageDTO,Image,ImageMetadata>
    {
       
        public Guid? Id {get; set;} = null;      

        public byte[] CurrentVersion { get; set; } = Array.Empty<byte>();

        public List<NoteSubDTO>? Notes {get; set;} = null;

        public Guid AlbumId {get; set;} = Guid.Empty;

        public bool Approved {get; set;} = false;

        public bool Visible {get; set;} = false;

        public List<ImageDTO> Media {get; set;} = new();

        public bool Locked {get; set;} = true;

        public string Name {get; set;} = string.Empty;

        public string Description {get; set;} = string.Empty;

        public Guid FranchiseId {get; set;} = Guid.Empty;

        public string Contact {get; set;} = string.Empty;

        public string FranchiseName {get; set;} = string.Empty;

        public long PriceMinor {get; set;} = 0;

        public bool Full {get; set;}=true;

        public ListingType Type  {get; set;} = ListingType.Generic;

        public ProductListingSubDTO? ProductListingExtension {get;set;} = null;
        public MedicalListingSubDTO? MedicalListingExtension {get; set;} = null;
        public AnimalListingSubDTO? AnimalListingExtension {get; set;} = null;

        public DiscountResponseSubDTO? ListingDiscount {get; set;} =null;

        public List<AvailabilityResponseSubDTO> Availability {get; set;} = new();

        public List<CommentResponseSubDTO> Comments {get; set;} = new();

        public DateTime Posted {get; set;} = DateTime.UtcNow;

        public string? MediaCreationToken {get; set;} = string.Empty;

        public static ListingResponseDTO? FromEntity(Listing? entity)
        {
            if(entity==null){return null;}
            ListingResponseDTO output = new ListingResponseDTO
            {
                Id = entity.Id,
                CurrentVersion=entity.CurrentVersion,
                Name=entity.ListingName,
                Description=entity.ListingDescription,
                FranchiseId=entity.FranchiseId,
                Contact="No provided contact.",
                FranchiseName="No provided name.",
                PriceMinor=entity.PriceMinor,
                Type=entity.Type,
                Approved=entity.Approved,
                Visible=entity.Visible,
                AlbumId=entity.AlbumId,
                ProductListingExtension=ProductListingSubDTO.FromEntity(entity.ProductExtension),
                MedicalListingExtension=MedicalListingSubDTO.FromEntity(entity.MedicalExtension),
                AnimalListingExtension=AnimalListingSubDTO.FromEntity(entity.AnimalExtension),
                ListingDiscount=DiscountResponseSubDTO.FromEntity(entity.ListingDiscount),
                Availability=entity.AvailabilityRecords.Select(a=>AvailabilityResponseSubDTO.FromEntity(a)!).ToList(),
                Comments=entity.Comments.Select(c=>CommentResponseSubDTO.FromEntity(c)!).ToList(),
                Posted=entity.Posted
            };

            if (entity.Business != null)
            {
                output.Contact=entity.Business.Contact??output.Contact;
                output.FranchiseName=entity.Business.FranchiseName??output.FranchiseName;
            }

            if (entity.Album != null)
            {
                output.Locked=entity.Album.Locked;
                output.Media = entity.Album.Images.Select(i=>ImageDTO.FromEntity(i)!).ToList();
                output.Full=entity.Album.Reserved>=entity.Album.Capacity;
            }

            return output;
        }


        public static ListingResponseDTO? FromEntity(Listing? entity, string token)
        {
            ListingResponseDTO? output = FromEntity(entity);

            if (output != null)
            {
                output.MediaCreationToken=token;
            }


            return output;
        }


    }
}

File: Frontend/pet_center_app/lib/models/data_transfer/listing/listing_response_dto.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:pet_center_app/models/data_transfer/listing/sub_dtos.dart';
import 'package:pet_center_app/models/data_transfer/note_sub_dto.dart';
import 'package:pet_center_app/models/data_transfer/image_dto.dart';
import 'package:pet_center_app/models/enums.dart';
part 'listing_response_dto.g.dart';

@JsonSerializable(explicitToJson: true)
class ListingResponseDTO {
  String? id;
  String currentVersion;
  List<NoteSubDTO>? notes;
  String? albumId;
  List<ImageDTO> media;
  bool locked;
  bool full;
  String name;
  String description;
  String franchiseId;
  String contact;
  String franchiseName;
  bool approved;
  bool visible;
  int priceMinor;
  ListingType type;
  ProductListingSubDTO? productListingExtension;
  MedicalListingSubDTO? medicalListingExtension;
  AnimalListingSubDTO? animalListingExtension;
  DiscountResponseSubDTO? listingDiscount;
  List<AvailabilityResponseSubDTO> availability;
  List<CommentResponseSubDTO> comments;
  String? mediaCreationToken;
  DateTime posted;

  ListingResponseDTO({
    this.id,
    this.currentVersion = '',
    this.notes,
    this.albumId,
    List<ImageDTO>? media,
    this.name = '',
    this.description = '',
    this.franchiseId = '',
    this.contact = '',
    this.franchiseName = '',
    this.locked = true,
    this.full = true,
    this.approved = false,
    this.visible = false,
    this.priceMinor = 0,
    DateTime? posted,
    this.type = ListingType.generic,
    this.productListingExtension,
    this.medicalListingExtension,
    this.animalListingExtension,
    this.listingDiscount,
    this.mediaCreationToken,
    List<AvailabilityResponseSubDTO>? availability,
    List<CommentResponseSubDTO>? comments,
  }) : media = media ?? [],
       availability = availability ?? [],
       comments = comments ?? [],
       posted = posted ?? DateTime.now().toUtc();

  factory ListingResponseDTO.fromJson(Map<String, dynamic> json) =>
      _$ListingResponseDTOFromJson(json);

  Map<String, dynamic> toJson() => _$ListingResponseDTOToJson(this);

  ListingResponseDTO copy() => ListingResponseDTO(
    id: id,
    currentVersion: currentVersion,
    notes: notes?.map((n) => n.copy()).toList(),
    albumId: albumId,
    media: media.map((m) => m.copy()).toList(),
    name: name,
    description: description,
    franchiseId: franchiseId,
    contact: contact,
    visible: visible,
    approved: approved,
    franchiseName: franchiseName,
    locked: locked,
    full: full,
    priceMinor: priceMinor,
    type: type,
    productListingExtension: productListingExtension?.copy(),
    medicalListingExtension: medicalListingExtension?.copy(),
    animalListingExtension: animalListingExtension?.copy(),
    listingDiscount: listingDiscount?.copy(),
    mediaCreationToken: mediaCreationToken,
    availability: availability.map((a) => a.copy()).toList(),
    comments: comments.map((c) => c.copy()).toList(),
  );
}
